### PR TITLE
[#16] Parse extensions that don't obey 'Read' instance

### LINKED ISSUE #16: Parse extensions that don't obey 'Read' instance

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #16

### CODE REVIEW

File: src/Extensions/Parser.hs [hunk 1]
```diff
@@ -111,4 +111,15 @@ cppP = [] <$ many newline <* try (char '#' <* noneOf "-") <* manyTill anyChar (t
...
 
+{- | Parse 'Extension' from a string. 'Read' instance for 'Extension'
+doesn't always work since some extensions are named differently.
+-}
+readExtension :: String -> Maybe Extension
+readExtension = \case
+    "GeneralisedNewtypeDeriving" -> Just GeneralizedNewtypeDeriving
+    "NamedFieldPuns" -> Just RecordPuns
+    "Rank2Types" -> Just RankNTypes
+    "CPP" -> Just Cpp
```
Comment (vrom911) [inline src/Extensions/Parser.hs:123]: Seems like `Cpp` is not working in module files.
Sorry, I am a bit confused in all these `Extension`s data types and their allowed scope (where they would work).

File: test/Test/Extensions/Parser.hs [hunk 2]
```diff
@@ -13,6 +13,7 @@ import Extensions.Parser (parseSource)
 parserSpec :: Spec
 parserSpec = describe "Haskell file Extensions Parser" $ do
     itShouldParse "{-# LANGUAGE TypeApplications #-}" [TypeApplications]
+    itShouldParse "{-# LaNgUaGe CPP #-}" [Cpp]
```
Comment (vrom911) [inline test/Test/Extensions/Parser.hs:16]: :snake:

File: src/Extensions/Parser.hs (hunk 1 above)
Comment (chshersh) [reply src/Extensions/Parser.hs:123]: Yes, this is indeed confusing... Cabal has its own type for extensions which is not the same as GHC extension type. Cabal works only with `.cabal` files and parses extensions from this file. And GHC only works with Haskell modules, it doesn't know anything about `.cabal` files. Problems happen when an extension is written in one way in a file (like `CPP`) but constructor name is different (`Cpp`), so `read` instance won't work. But this means that I also need to handle `Cpp` string as `Nothing`...

**Review by vrom911 — APPROVED** [review]
Thanks for clarifications! Now it makes more sense :+1:

**Pull request merged**